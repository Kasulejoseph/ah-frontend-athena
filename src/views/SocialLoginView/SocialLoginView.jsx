import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router";
import PropTypes from "prop-types";
import { toast } from "react-toastify";
import SocialLoginButtons from "../../components/SocialLogin/SocialLogin";
import googleLoginActionFunction from "../../actions/googleLoginAction";
import facebookLoginActionFunction from "../../actions/facebookLoginAction";

export class SocialLoginView extends Component {
  /**
   *
   * @param {} props - props passed into the JSX initialisation of the component
   */
  constructor(props) {
    super(props);
    this.state = {};
    this.handleGoogleLogin = this.handleGoogleLogin.bind(this);
    this.handleFacebookLogin = this.handleFacebookLogin.bind(this);
  }

  /**
   * Receives props after action is dispatched.
   * @param {} nextProps - properties object that is updated when an
   * action is dispatched and mapStateToProps below is invoked.
   */
  componentWillReceiveProps(nextProps) {
    let googleResponse = nextProps.returnedDataGoogle;
    if (googleResponse) {
      let googleError = googleResponse.errors;
      let googleUser = googleResponse.user;

      if (googleError) {
        toast.error(googleError.auth_token[0]);
      } else if (googleUser) {
        let jwt_token = googleUser.jwt_token;
        if (jwt_token) {
          toast.success("Successfully logged in with Google.");
          const { history } = this.props;
          window.localStorage.setItem("token", jwt_token);
          history.push("/");
        }
      } else {
        toast.error("Error in Google Auth system. Please try again.");
      }
    }

    let facebookResponse = nextProps.returnedDataFacebook;
    if (facebookResponse) {
      let facebookErrors = facebookResponse.errors;
      let facebookUser = facebookResponse.user;
      if (facebookErrors) {
        toast.error(facebookErrors.auth_token[0]);
      } else if (facebookUser) {
        let jwt_token = facebookUser.jwt_token;
        if (jwt_token) {
          toast.success("Successfully logged in with Facebook.");
          const { history } = this.props;
          window.localStorage.setItem("token", jwt_token);
          history.push("/");
        }
      } else {
        toast.error("Error in Facebook Auth system. Please try again.");
      }
    }
  }

  /**
   * This function is triggered when the user clicks on the signin with Google button.
   * @param {} response - response object from google signin (user data & tokenId)
   */
  handleGoogleLogin = response => {
    if (response.tokenId) {
      const { googleLoginActionFunction } = this.props;
      googleLoginActionFunction(response.tokenId);
    }
  };

  /**
   * This function is triggered when the user clicks on the signin with Facebook button.
   * @param {} response - response object from facebook signin (user data & accessToken)
   */
  handleFacebookLogin = response => {
    if (response.accessToken) {
      const { facebookLoginActionFunction } = this.props;
      facebookLoginActionFunction(response.accessToken);
    }
  };

  render() {
    return (
      <SocialLoginButtons
        facebookLoginHandler={this.handleFacebookLogin}
        googleLoginHandler={this.handleGoogleLogin}
      />
    );
  }
}

SocialLoginView.propTypes = {
  history: PropTypes.shape({}),
  facebookLoginActionFunction: PropTypes.func.isRequired,
  googleLoginActionFunction: PropTypes.func.isRequired,
  returnedDataGoogle: PropTypes.shape({}),
  returnedDataFacebook: PropTypes.shape({})
};
SocialLoginView.defaultProps = {
  history: {},
  returnedDataGoogle: undefined,
  returnedDataFacebook: undefined
};

/**
 * Filters which state values are mapped to current component's props.
 * @param {} state - state object passed down from the root reducer
 */
export const mapStateToProps = state => ({
  returnedDataGoogle: state.googleLogin.responseData,
  returnedDataFacebook: state.facebookLogin.responseData
});

export default withRouter(
  connect(
    mapStateToProps,
    { googleLoginActionFunction, facebookLoginActionFunction }
  )(SocialLoginView)
);
