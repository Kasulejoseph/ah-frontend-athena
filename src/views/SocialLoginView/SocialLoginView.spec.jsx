import React from "react";
import { shallow } from "enzyme";
import { SocialLoginView, mapStateToProps } from "./SocialLoginView";

let push = jest.fn();
const googleLoginActionFunction = jest.fn();
const facebookLoginActionFunction = jest.fn();

let props = {
  googleLoginActionFunction,
  facebookLoginActionFunction,
  googleLogin: {
    responseData: {
      user: { jwt_token: "mocked token" },
      errors: { auth_token: ["faked facebook login error"] }
    }
  },
  facebookLogin: {
    responseData: {
      user: { jwt_token: "mocked token" },
      errors: { auth_token: ["faked facebook login error"] }
    }
  },
  history: { push }
};

const initialState = {
  googleLogin: {
    responseData: {
      user: { jwt_token: "mocked token" },
      errors: { auth_token: ["faked facebook login error"] }
    }
  },
  facebookLogin: {
    responseData: {
      user: { jwt_token: "mocked token" },
      errors: { auth_token: ["faked facebook login error"] }
    }
  }
};
let wrapper;

describe("Test SocialLoginView stateful component", () => {
  beforeEach(() => {
    wrapper = shallow(<SocialLoginView {...props} />);
  });

  it("should match the SocialLoginView snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should receive and handle nextProps properly", () => {
    wrapper.setProps({
      returnedDataFacebook: { user: { jwt_token: "mocked token" } }
    });
    expect(wrapper.instance().props.returnedDataFacebook).toEqual({
      user: { jwt_token: "mocked token" }
    });
  });

  it("should receive and handle facebook error data properly", () => {
    wrapper.setProps({
      returnedDataFacebook: {
        errors: { auth_token: ["faked facebook login error"] }
      }
    });
    expect(wrapper.instance().props.returnedDataFacebook).toEqual({
      errors: { auth_token: ["faked facebook login error"] }
    });
  });

  it("should handle facebook fatal error data properly", () => {
    wrapper.setProps({
      returnedDataFacebook: undefined
    });
    expect(wrapper.instance().props.returnedDataFacebook).toEqual(undefined);
  });

  it("should receive and handle google success properly", () => {
    wrapper.setProps({
      returnedDataGoogle: { user: { jwt_token: "mocked token" } }
    });
    expect(wrapper.instance().props.returnedDataGoogle).toEqual({
      user: { jwt_token: "mocked token" }
    });
  });

  it("should receive and handle google error data properly", () => {
    wrapper.setProps({
      returnedDataGoogle: {
        errors: { auth_token: ["faked google login error"] }
      }
    });
    expect(wrapper.instance().props.returnedDataGoogle).toEqual({
      errors: { auth_token: ["faked google login error"] }
    });
  });

  it("should handle facebook fatal error data properly", () => {
    wrapper.setProps({
      returnedDataFacebook: undefined
    });
    expect(wrapper.instance().props.returnedDataFacebook).toEqual(undefined);
  });

  it("should map state to props", () => {
    expect(mapStateToProps(initialState).returnedDataFacebook["user"]).toEqual({
      jwt_token: "mocked token"
    });
  });

  it("should invoke handleGoogleLogin function and call action", () => {
    wrapper.instance().handleGoogleLogin({
      tokenId: "mocked token id from google"
    });
    expect(googleLoginActionFunction).toBeCalled();
  });

  it("should invoke handleFacebookLogin function and call action", () => {
    wrapper.instance().handleFacebookLogin({
      accessToken: "mocked token id from facebook"
    });
    expect(googleLoginActionFunction).toBeCalled();
  });
});
