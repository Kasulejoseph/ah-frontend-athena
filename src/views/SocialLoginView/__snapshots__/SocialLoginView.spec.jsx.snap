// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SocialLoginView stateful component should match the SocialLoginView snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SocialLoginView
    facebookLogin={
      Object {
        "responseData": Object {
          "errors": Object {
            "auth_token": Array [
              "faked facebook login error",
            ],
          },
          "user": Object {
            "jwt_token": "mocked token",
          },
        },
      }
    }
    facebookLoginActionFunction={[MockFunction]}
    googleLogin={
      Object {
        "responseData": Object {
          "errors": Object {
            "auth_token": Array [
              "faked facebook login error",
            ],
          },
          "user": Object {
            "jwt_token": "mocked token",
          },
        },
      }
    }
    googleLoginActionFunction={[MockFunction]}
    history={
      Object {
        "push": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "facebookLoginHandler": [Function],
      "googleLoginHandler": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "facebookLoginHandler": [Function],
        "googleLoginHandler": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
